src/audio/mux.c

```diff
@@ -26,52 +26,366 @@
+			continue;
+		sinks_bytes[i] = frames * comp_frame_bytes(sinks[i]->sink);
+	}
+
+	/* produce output, one sink at a time */
+	for (i = 0; i < MUX_MAX_STREAMS; i++) {
+		if (!sinks[i])
+			continue;
+
+		cd->mux(dev, sinks[i], source, frames, &cd->streams[i]);
+	}
+
+	/* update components */
+	for (i = 0; i < MUX_MAX_STREAMS; i++) {
+		if (!sinks[i])
+			continue;
+		comp_update_buffer_produce(sinks[i], sinks_bytes[i]);
+	}
+	comp_update_buffer_consume(source, source_bytes);
+
+	return 0;
+}
+
+static inline int demux_process(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *sink;
+	struct comp_buffer *source;
+	struct comp_buffer *sources[MUX_MAX_STREAMS] = { NULL };
+	struct list_item *clist;
+	uint32_t num_sources = 0;
+	uint32_t i = 0;
+	uint32_t frames = -1;
+	uint32_t sources_bytes[MUX_MAX_STREAMS] = { 0 };
+	uint32_t sink_bytes;
+
+	tracev_mux("demux_process()");
+
+	// align source streams with their respective configurations
+	list_for_item(clist, &dev->bsource_list) {
+		source = container_of(clist, struct comp_buffer, sink_list);
+		if (source->source->state == dev->state) {
+			num_sources++;
+			i = get_stream_index(cd, source->ipc_buffer.comp.pipeline_id);
+			sources[i] = source;
+		}
+	}
+
+	/* check if there are any sources active */
+	if (num_sources == 0)
+		return 0;
+
+	sink = list_first_item(&dev->bsink_list, struct comp_buffer,
+			       source_list);
+
+	/* check if sink is active */
+	if (sink->sink->state != dev->state)
+		return 0;
+
+	/* check for underrun */
+	for (i = 0; i < MUX_MAX_STREAMS; i++) {
+		if (!sources[i])
+			continue;
+		if (sources[i]->avail == 0) {
+			trace_mux_error("demux_process() error: source "
+					"component buffer has not enough data "
+					"avaialble.");
+			comp_underrun(dev, sources[i], 0, 0);
+			return -EIO;
+		}
+	}
+
+	/* check for overrun */
+	if (sink->free == 0) {
+		trace_mux_error("demux_process() error: sink component "
+				"buffer has not enough free bytes.");
+		comp_overrun(dev, sink, 0, 0);
+		return -EIO;
+	}
+
+	for (i = 0; i < MUX_MAX_STREAMS; i++) {
+		if (!sources[i])
+			continue;
+		frames = MIN(frames, comp_avail_frames(sources[i], sink));
+	}
+
+	for (i = 0; i < MUX_MAX_STREAMS; i++) {
+		if (!sources[i])
+			continue;
+		sources_bytes[i] = frames *
+				   comp_frame_bytes(sources[i]->source);
+	}
+	sink_bytes = frames * comp_frame_bytes(sink->sink);
+
+	/* produce output */
+	cd->demux(dev, sink, &sources[0], frames, &cd->streams[0]);
+
+	/* update components */
+	comp_update_buffer_produce(sink, sink_bytes);
+	for (i = 0; i < MUX_MAX_STREAMS; i++) {
+		if (!sources[i])
+			continue;
+		comp_update_buffer_consume(sources[i], sources_bytes[i]);
+	}
+
 	return 0;
 }
 
 /* copy and process stream data from source to sink buffers */
 static int mux_copy(struct comp_dev *dev)
 {
+	struct comp_data *cd = comp_get_drvdata(dev);
+	int ret = 0;
 
-	return 0;
+	tracev_mux("mux_copy()");
+
+	if (cd->mode == MODE_MUX)
```

Comment:
Instead of handling the component mode (with a bit inconsistent types as commented in the header file) all the time at run-time and doing branching/redirection every frame, each mode could be registered as a dedicated comp_driver (all common ops pointed by both), with a dedicated component id and be selected in the topology by the comp id, not by the configured mode. How about this idea?

Reply:
I think that this is a great idea, but it also conflicts with @ranj063 comments, about unification of these functions as being mirrored.
The fact is, that during run-time, mode of this component is 'fixed' by rigid structure in the topology, so it is very unlikely to ever be changed in run-time.
This is very much an optimization between code size and run-time performance and I believe, that gain in performance is more significant in this case.